File: internal/vault/static/behavior.js
/**
 * Behavioral biometrics for the card input (§14.2 extension).
 *
 * How a card number is entered separates a cardholder from a fraudster
 * surprisingly well, and this iframe is the only place in the entire system
 * that can observe it — the merchant's page cannot reach inside, and by the
 * time the charge reaches the API the keystrokes are long gone.
 *
 * The signals, roughly in order of how much they are worth:
 *
 *   PASTE           A cardholder types the number off the card in their hand.
 *                   A fraudster pastes it from a list of stolen numbers. This
 *                   is the single most discriminating signal here and it costs
 *                   one event listener.
 *
 *   RHYTHM          Humans type unevenly — digits cluster in groups of four
 *                   with pauses between. Scripted input is metronomic. Low
 *                   inter-keystroke variance means automation.
 *
 *   CORRECTIONS     You rarely mistype your own card. You often mistype one
 *                   you are reading off another screen.
 *
 *   CVV HESITATION  A cardholder knows their CVV or reads it off the back in
 *                   one motion. Someone hunting for it in a dump pauses.
 *
 *   TOTAL DURATION  Both extremes are suspicious: implausibly fast means
 *                   automation, very slow means unfamiliarity.
 *
 * PRIVACY — the constraint that shapes this file.
 *
 * Keystroke dynamics are biometric data under GDPR Art. 9 and Illinois BIPA.
 * So nothing here records WHAT was typed, and nothing records a per-keystroke
 * timeline that could be replayed to reconstruct it. Only aggregates leave
 * this frame: counts, a mean, a variance, a coefficient of variation. Those
 * cannot be inverted into a card number, and they are not personal data on
 * their own.
 *
 * A per-keystroke timing array WOULD be more predictive. It is deliberately
 * not collected.
 */

export function createBehaviorTracker() {
  const state = {
    fields: {},
    startedAt: null,
    pasteCount: 0,
    pastedFields: [],
  };

  function field(name) {
    if (!state.fields[name]) {
      state.fields[name] = {
        keystrokes: 0,
        corrections: 0,
        intervals: [],
        focusedAt: null,
        firstKeyAt: null,
        lastKeyAt: null,
        pasted: false,
      };
    }
    return state.fields[name];
  }

  function attach(el, name) {
    const f = field(name);

    el.addEventListener('focus', () => {
      f.focusedAt = performance.now();
      if (state.startedAt === null) state.startedAt = f.focusedAt;
    });

    el.addEventListener('keydown', (e) => {
      const now = performance.now();

      // Backspace and Delete are corrections. Tracked as a rate rather than a
      // position, so nothing reveals which digit was wrong.
      if (e.key === 'Backspace' || e.key === 'Delete') {
        f.corrections++;
        return;
      }
      // Only count character input. Modifiers and navigation are not typing.
      if (e.key.length !== 1) return;

      f.keystrokes++;
      if (f.firstKeyAt === null) f.firstKeyAt = now;

      if (f.lastKeyAt !== null) {
        const gap = now - f.lastKeyAt;
        // Gaps over 5s are the user looking away, not typing rhythm. Including
        // them would swamp the variance with a single interruption.
        if (gap < 5000) f.intervals.push(gap);
      }
      f.lastKeyAt = now;
    });

    el.addEventListener('paste', () => {
      f.pasted = true;
      state.pasteCount++;
      if (!state.pastedFields.includes(name)) state.pastedFields.push(name);
    });
  }

  function stats(intervals) {
    if (intervals.length < 2) return { mean: null, cv: null };

    const mean = intervals.reduce((a, b) => a + b, 0) / intervals.length;
    const variance =
      intervals.reduce((sum, v) => sum + (v - mean) ** 2, 0) / intervals.length;
    const stddev = Math.sqrt(variance);

    // Coefficient of variation, not raw stddev: it is scale-free, so a fast
    // typist and a slow one with the same rhythm score alike. What separates
    // human from script is the EVENNESS, not the speed.
    return { mean: round(mean), cv: mean > 0 ? round(stddev / mean, 4) : null };
  }

  function round(n, places = 1) {
    const f = Math.pow(10, places);
    return Math.round(n * f) / f;
  }

  /**
   * Summarize into the aggregates that leave this frame.
   * Nothing here can be inverted into the entered values.
   */
  function summarize() {
    const number = field('number');
    const cvc = field('cvc');
    const numberStats = stats(number.intervals);

    // Time from first keystroke to last, across all fields.
    const ends = Object.values(state.fields)
      .map((f) => f.lastKeyAt)
      .filter((t) => t !== null);
    const totalMs =
      state.startedAt !== null && ends.length
        ? round(Math.max(...ends) - state.startedAt)
        : null;

    return {
      // The headline signal.
      pasted: state.pasteCount > 0,
      pasted_fields: state.pastedFields,

      // Rhythm. cv near 0 means metronomic, i.e. scripted.
      keystroke_interval_mean_ms: numberStats.mean,
      keystroke_interval_cv: numberStats.cv,

      // Effort. A high correction rate suggests reading an unfamiliar number.
      number_keystrokes: number.keystrokes,
      number_corrections: number.corrections,
      correction_rate:
        number.keystrokes > 0
          ? round(number.corrections / number.keystrokes, 3)
          : null,

      // Hesitation before the CVV: knowing it vs hunting for it.
      cvc_hesitation_ms:
        cvc.focusedAt !== null && cvc.firstKeyAt !== null
          ? round(cvc.firstKeyAt - cvc.focusedAt)
          : null,

      total_duration_ms: totalMs,
    };
  }

  return { attach, summarize };
}

File: internal/risk/behavior.go
package risk

// Behavioral biometrics rules (§14.2 extension).
//
// These are RULES, not model features, and that is not a shortcut. The training
// data has no keystroke information at all, so there is no weight to learn —
// any number the model produced for these would be invented. Encoding them as
// explicit, explainable rules is the honest option, and it is exactly what the
// rule engine exists for: policy we can state and defend, sitting underneath a
// model that cannot override it.
//
// When real traffic accumulates, chargeback outcomes become labels for these
// signals and they can graduate into the model (§14.3). Until then they stay
// here.
//
// A deliberate calibration note: nearly all of these CHALLENGE rather than
// BLOCK. Pasting a card number is common and legitimate — password managers do
// it, and so does anyone who keeps their card in a notes app. Blocking on paste
// alone would decline a large number of real customers to catch a few
// fraudsters, which is the wrong trade for almost every merchant. The one
// exception is a paste combined with machine-like rhythm, which no human
// produces.

// Behavior carries the aggregates collected by the checkout iframe.
//
// Note what is absent: any per-keystroke timing array, and anything that
// records what was typed. Keystroke dynamics are biometric data under GDPR
// Art. 9 and Illinois BIPA, so only non-invertible aggregates cross the
// boundary. A full timing sequence would predict better and is deliberately
// not collected.
type Behavior struct {
	// Present reports whether the iframe supplied any data. Old widget
	// versions and non-browser API callers will not, and their absence must
	// not be read as suspicious — see the guard on every rule below.
	Present bool

	Pasted       bool
	PastedFields []string

	// KeystrokeIntervalCV is the coefficient of variation of inter-keystroke
	// gaps. Scale-free on purpose: a fast typist and a slow one with the same
	// rhythm score alike. Human typing is uneven (CV typically 0.4-1.2);
	// scripted input is metronomic (CV near 0).
	KeystrokeIntervalCV   *float64
	KeystrokeIntervalMean *float64

	NumberKeystrokes int
	CorrectionRate   *float64
	CVCHesitationMs  *float64
	TotalDurationMs  *float64
}

// Thresholds, added to Config.
type BehaviorConfig struct {
	// Below this CV, typing is too even to be human.
	MinKeystrokeCV float64 `yaml:"min_keystroke_cv"`
	// Above this correction rate, the typist is likely reading an unfamiliar
	// number rather than one they own.
	MaxCorrectionRate float64 `yaml:"max_correction_rate"`
	// A long pause before the CVV suggests hunting for it.
	MaxCVCHesitationMs float64 `yaml:"max_cvc_hesitation_ms"`
	// Faster than a person can physically type a 16-digit number.
	MinTotalDurationMs float64 `yaml:"min_total_duration_ms"`
}

func DefaultBehaviorConfig() BehaviorConfig {
	return BehaviorConfig{
		// Human inter-keystroke CV is rarely under ~0.25 even for practised
		// typists; digits cluster in groups with pauses between.
		MinKeystrokeCV: 0.15,
		// Mistyping a fifth of the digits of a card you are holding is
		// implausible.
		MaxCorrectionRate: 0.20,
		// Eight seconds staring at a CVV field you have already focused.
		MaxCVCHesitationMs: 8000,
		// ~1.2s is about the floor for a human entering 16 digits plus expiry
		// and CVV. Below that is automation, not a fast typist.
		MinTotalDurationMs: 1200,
	}
}

func behaviorRules() []Rule {
	return []Rule{
		{
			// Machine-paced typing. The clearest automation signal in the set:
			// humans do not type at a constant interval, and the CV is
			// scale-free so this does not simply flag fast typists.
			Name:        "automation_typing_rhythm",
			Description: "Keystroke timing is too regular to be human",
			Action:      ActionChallenge,
			Enabled:     true,
			Evaluate: func(t Transaction, c *Config) bool {
				b := t.Behavior
				// Needs enough keystrokes for the statistic to mean anything.
				if !b.Present || b.KeystrokeIntervalCV == nil || b.NumberKeystrokes < 8 {
					return false
				}
				return *b.KeystrokeIntervalCV < c.Behavior.MinKeystrokeCV
			},
		},
		{
			// Paste AND machine rhythm together. Either alone is weak; both at
			// once is a script filling a form, which no legitimate customer
			// produces.
			Name:        "automated_card_entry",
			Description: "Card details were pasted and entered at machine speed",
			Action:      ActionBlock,
			Enabled:     true,
			Evaluate: func(t Transaction, c *Config) bool {
				b := t.Behavior
				if !b.Present || !b.Pasted {
					return false
				}
				tooFast := b.TotalDurationMs != nil && *b.TotalDurationMs < c.Behavior.MinTotalDurationMs
				tooEven := b.KeystrokeIntervalCV != nil && b.NumberKeystrokes >= 8 &&
					*b.KeystrokeIntervalCV < c.Behavior.MinKeystrokeCV
				return tooFast || tooEven
			},
		},
		{
			// Paste on its own. Weighted low deliberately: password managers
			// paste, and plenty of people keep a card in a notes app. It is a
			// signal, not a verdict.
			Name:        "card_number_pasted",
			Description: "Card number was pasted rather than typed",
			Action:      ActionScore,
			Weight:      15,
			Enabled:     true,
			Evaluate: func(t Transaction, _ *Config) bool {
				return t.Behavior.Present && t.Behavior.Pasted
			},
		},
		{
			// You rarely mistype the card in your hand. You often mistype one
			// you are reading off another screen.
			Name:        "high_correction_rate",
			Description: "Card number was corrected unusually often while typing",
			Action:      ActionScore,
			Weight:      15,
			Enabled:     true,
			Evaluate: func(t Transaction, c *Config) bool {
				b := t.Behavior
				if !b.Present || b.CorrectionRate == nil || b.NumberKeystrokes < 8 {
					return false
				}
				return *b.CorrectionRate > c.Behavior.MaxCorrectionRate
			},
		},
		{
			// A cardholder knows their CVV or turns the card over. Hunting for
			// it in a list takes noticeably longer.
			Name:        "cvc_hesitation",
			Description: "Long pause before entering the security code",
			Action:      ActionScore,
			Weight:      10,
			Enabled:     true,
			Evaluate: func(t Transaction, c *Config) bool {
				b := t.Behavior
				if !b.Present || b.CVCHesitationMs == nil {
					return false
				}
				return *b.CVCHesitationMs > c.Behavior.MaxCVCHesitationMs
			},
		},
		{
			// Faster than a person can physically enter the fields, with no
			// paste to explain it — a script driving the DOM directly.
			Name:        "impossible_entry_speed",
			Description: "Card details were entered faster than a person can type",
			Action:      ActionChallenge,
			Enabled:     true,
			Evaluate: func(t Transaction, c *Config) bool {
				b := t.Behavior
				if !b.Present || b.TotalDurationMs == nil || b.Pasted {
					return false
				}
				// Requires real keystrokes: a legitimately autofilled form has
				// a short duration and no keystrokes, and must not be flagged.
				return b.NumberKeystrokes >= 8 && *b.TotalDurationMs < c.Behavior.MinTotalDurationMs
			},
		},
	}
}
